validator/server_api/validation_client.go

```diff
@@ -96,14 +98,66 @@ func (c *ValidationClient) Room() int {
 	return int(room32)
 }
 
+type GetLeavesHashRequest struct {
+	ExecutionID       uint64
+	FromBatch         uint64
+	MachineStartIndex uint64
+	StepSize          uint64
+	NumDesiredLeaves  uint64
+}
+
+// Generic marshaller
+type jsonMarshaller[T any] struct{}
+
+// Marshal converts a GetLeavesHashRequest into a JSON byte slice.
+func (j jsonMarshaller[T]) Marshal(v T) []byte {
+	data, err := json.Marshal(v)
+	if err != nil {
+		log.Error("error marshaling", "value", v, "error", err)
+		return nil
+	}
+	return data
+}
+
+// Unmarshal converts a JSON byte slice into a GetLeavesHashRequest.
+func (j jsonMarshaller[T]) Unmarshal(val []byte) (T, error) {
+	var v T
+	err := json.Unmarshal(val, &v)
+	if err != nil {
+		return v, err
+	}
+	return v, nil
+}
+
 type ExecutionClient struct {
 	ValidationClient
+
+	producer *pubsub.Producer[GetLeavesHashRequest, []common.Hash]
+}
+
+type RedisStreamCfg struct {
+	producerCfg *pubsub.ProducerConfig
+}
+
+type ExecutionClientOpts struct {
```

Comment:
config structs should only contain config values. Not fetchers and certainly not node/stack/et.
You want to have:
* ExecutionClientConfig struct would include ProducerConfig and a ClientConfig
* ExecutionClientConfigFetcher which is just a function returning ExecutionClientConfig
* NewExecutionClient recieves ExecutionClientConfigFetcher and *node.Node
* * The function can create a ClientConfigFetcher from it's ExecutionClientConfigFetcher

Reply:
Not a config struct, that's an option struct (https://google.github.io/styleguide/go/best-practices#option-structure), but that part wasn't meant for a review, dropped it altogether to draw your attention to review using pubsub part only.